systems/physics/src/lib.rs

```diff
@@ -12,52 +12,117 @@
         _ => Err("bad dispatch".into()),
     }
 }
 
-/// Receives messages on the subject `systems.{shard}.{system}.frames`, e.g. `systems.default.physics.frames`
-fn handle_frame(
+/// Routes message either to the `handle_ping` function for registry pings or `handle_frame` for position updates
+fn handle_message(
     ctx: &CapabilitiesContext,
     msg: impl Into<messaging::DeliverMessage>,
 ) -> CallResult {
     let msg = msg.into().message;
+    let subject = msg
+        .as_ref()
+        .map_or(NO_MESSAGE.to_string(), |m| m.subject.to_string());
     ctx.log(&format!(
         "Received message from broker on subject '{}'",
-        msg.as_ref()
-            .map_or("(no message)".to_string(), |m| m.subject.to_string())
+        subject
     ));
-    let bodyvec = msg.unwrap().body;
+    match subject.as_ref() {
+        NO_MESSAGE => Err("No message".into()),
+        REGISTRY_SUBJECT => handle_ping(ctx, msg.unwrap()),
+        _ => handle_frame(ctx, msg.unwrap()),
+    }
+}
+
+/// Receives messages on the subject `system.registry` and replies with physics system metadata
+fn handle_ping(
+    ctx: &CapabilitiesContext,
+    msg: guest::prelude::messaging::BrokerMessage,
+) -> CallResult {
+    let payload = System {
+        name: "physics".to_string(),
+        framerate: 60,
+        components: vec!["Position".to_string(), "Velocity".to_string()],
+    };
+    let ref reply_to = match msg.reply_to.len() {
+        0 => format!("decs.{}", REGISTRY_SUBJECT),
+        _ => msg.reply_to,
+    };
+    if let Err(_) = ctx
+        .msg()
+        .publish(reply_to, None, &serde_json::to_vec(&payload).unwrap())
+    {
+        return Err("Error publishing message".into());
+    };
+    Ok(vec![])
+}
 
-    // This frame should have three objects: "elapsed", "position" and "velocity"
-    let (_elapsed, _pos, _vel) = extract_frame(&bodyvec)?;
+/// Receives messages on the subject `decs.systems.{shard}.{system}.frames`, e.g. `decs.systems.default.physics.frames`
+fn handle_frame(
+    ctx: &CapabilitiesContext,
+    msg: guest::prelude::messaging::BrokerMessage,
+) -> CallResult {
+    let subject: Vec<&str> = msg.subject.split(".").collect();
+    if subject.len() != 5 {
+        return Err("Unknown message subject received".into());
+    }
 
+    let data = extract_frame(&msg.body);
+    if let Err(_) = data {
+        return Err("Did not receive components needed for frame update".into());
+    }
+
+    let (entity, elapsed, pos, vel) = data.unwrap();
+    if vel.mag == 0 {
+        return Ok(vec![]);
+    } else if vel.ux == 0.0 && vel.uy == 0.0 && vel.uz == 0.0 {
+        return Err("Bad target vector".into());
+    }
+
+    if let Ok(new_position) = new_position(elapsed, &pos, &vel) {
+        let publish_subject = &format!("decs.{}.{}.position.put", subject[2], entity);
+        if let Err(_) = ctx.msg().publish(
+            publish_subject,
+            None,
+            &serde_json::to_vec(&new_position).unwrap(),
```

Comment:
since the return type of this function is `CallResult`, you can use the `?` syntax to avoid unwrapping... you get the real value, or you get an error and the function returns from the call site. So here you could just use `&serde_json::to_vec(&new_position)?,`

Reply:
Removed unwrap and replaced with ?